src/Microsoft.ManagedZlib/src/DeflateStream.cs

```diff
@@ -0,0 +1,1078 @@
+                _stream = stream;
+                _mode = CompressionMode.Decompress;
+                _leaveOpen = leaveOpen;
+                break;
+
+            case CompressionMode.Compress:
+                InitializeDeflater(stream, leaveOpen, windowBits, CompressionLevel.Optimal);
+                break;
+
+            default:
+                throw new ArgumentException("ArgumentOutOfRange_Enum - Enum value was out of legal range.", nameof(mode));
+        }
+    }
+
+    /// <summary>
+    /// Internal constructor to specify the compressionlevel as well as the windowbits
+    /// </summary>
+    internal DeflateStream(Stream stream, CompressionLevel compressionLevel, bool leaveOpen, int windowBits)
+    {
+        ArgumentNullException.ThrowIfNull(stream);
+
+        InitializeDeflater(stream, leaveOpen, windowBits, compressionLevel);
+    }
+
+    /// <summary>
+    /// Sets up this DeflateStream to be used for Zlib Deflation/Compression
+    /// </summary>
+    [MemberNotNull(nameof(_stream))]
+    internal void InitializeDeflater(Stream stream, bool leaveOpen, int windowBits, CompressionLevel compressionLevel)
+    {
+        Debug.Assert(stream != null);
+        if (!stream.CanWrite)
+            throw new ArgumentException("NotSupported_UnwritableStream - Stream does not support writing.", nameof(stream));
+
+        _deflater = new Deflater(compressionLevel, windowBits);
+
+        _stream = stream;
+        _mode = CompressionMode.Compress;
+        _leaveOpen = leaveOpen;
+        InitializeBuffer();
+    }
+
+    [MemberNotNull(nameof(_buffer))]
+    private void InitializeBuffer()
+    {
+        Debug.Assert(_buffer == null);
+        _buffer = ArrayPool<byte>.Shared.Rent(DefaultBufferSize);
+    }
+
+    [MemberNotNull(nameof(_buffer))]
+    private void EnsureBufferInitialized()
+    {
+        if (_buffer == null)
+        {
+            InitializeBuffer();
+        }
+    }
+
+    public Stream BaseStream => _stream;
+
+    public override bool CanRead
+    {
+        get
+        {
+            if (_stream == null)
+            {
+                return false;
+            }
+
+            return (_mode == CompressionMode.Decompress && _stream.CanRead);
+        }
+    }
+
+    public override bool CanWrite
+    {
+        get
+        {
+            if (_stream == null)
+            {
+                return false;
+            }
+
+            return (_mode == CompressionMode.Compress && _stream.CanWrite);
+        }
+    }
+
+    public override bool CanSeek => false;
+
+    public override long Length
+    {
+        get { throw new NotSupportedException("NotSupported - This operation is not supported."); }
+    }
+
+    public override long Position
+    {
+        get { throw new NotSupportedException("NotSupported - This operation is not supported."); }
+        set { throw new NotSupportedException("NotSupported - This operation is not supported."); }
+    }
```

Comment:
1. The strings are redundant. The exception class name is sufficiently descriptive.
2. IDE0027: Use expression body for accessors.

```suggestion
    public override long Length => throw new NotSupportedException();

    public override long Position
    {
        get => throw new NotSupportedException();
        set => throw new NotSupportedException();
    }
```